File: BAEKJOON/2239 - 스도쿠/스도쿠.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        // 입력 받기
        var board = new int[9, 9];
        for (int i = 0; i < 9; i++)
        {
            string input = Console.ReadLine();
            for (int j = 0; j < 9; j++)
            {
                board[i, j] = input[j] - '0';
            }
        }

        // 스도쿠 풀기
        SolveSudoku(board);

        // 결과 출력
        PrintBoard(board);
    }

    static bool IsValid(int[,] board, int row, int col, int num)
    {
        // 해당 행에 num이 있는지 확인
        for (int i = 0; i < 9; i++)
        {
            if (board[row, i] == num)
            {
                return false;
            }
        }

        // 해당 열에 num이 있는지 확인
        for (int i = 0; i < 9; i++)
        {
            if (board[i, col] == num)
            {
                return false;
            }
        }

        // 3x3 박스 내에 num이 있는지 확인
        int startRow = 3 * (row / 3);
        int startCol = 3 * (col / 3);
        for (int i = 0; i < 3; i++)
        {
            for (int j = 0; j < 3; j++)
            {
                if (board[startRow + i, startCol + j] == num)
                {
                    return false;
                }
            }
        }

        // num을 놓을 수 있으면 true 반환
        return true;
    }

    static bool Backtrack(int[,] board, List<(int, int)> emptyCells, int index)
    {
        // 모든 빈 칸을 다 채웠다면 true 반환
        if (index == emptyCells.Count)
        {
            return true;
        }

        // 현재 빈 칸의 좌표 가져오기
        var (row, col) = emptyCells[index];

        // 1부터 9까지의 숫자 시도
        for (int num = 1; num <= 9; num++)
        {
            if (IsValid(board, row, col, num))
            {
                // 유효한 숫자라면 보드에 숫자 배치
                board[row, col] = num;

                // 다음 빈 칸으로 이동
                if (Backtrack(board, emptyCells, index + 1))
                {
                    return true;
                }

                // 현재 숫자가 유효하지 않다면 다시 0으로 되돌림
                board[row, col] = 0;
            }
        }

        // 가능한 숫자가 없다면 false 반환
        return false;
    }

    static bool SolveSudoku(int[,] board)
    {
        // 빈 칸들의 좌표를 리스트로 저장
        var emptyCells = new List<(int, int)>();
        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                if (board[i, j] == 0)
                {
                    emptyCells.Add((i, j));
                }
            }
        }

        // 백트래킹 시작
        return Backtrack(board, emptyCells, 0);
    }

    static void PrintBoard(int[,] board)
    {
        // 보드 출력
        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                Console.Write(board[i, j]);
            }
            Console.WriteLine();
        }
    }
}

File: BAEKJOON/2239 - 스도쿠/스도쿠.py
def is_valid(board, row, col, num):
    # 해당 행에 num이 있는지 확인
    for i in range(9):
        if board[row][i] == num:
            return False
    
    # 해당 열에 num이 있는지 확인
    for i in range(9):
        if board[i][col] == num:
            return False
    
    # 3x3 박스 내에 num이 있는지 확인
    start_row, start_col = 3 * (row // 3), 3 * (col // 3)
    for i in range(3):
        for j in range(3):
            if board[start_row + i][start_col + j] == num:
                return False
    
    # num을 놓을 수 있으면 True 반환
    return True

def solve_sudoku(board):
    # 빈 칸들의 좌표를 리스트로 저장
    empty_cells = [(i, j) for i in range(9) for j in range(9) if board[i][j] == 0]
    
    # 백트래킹 시작
    return backtrack(board, empty_cells, 0)

def backtrack(board, empty_cells, index):
    # 모든 빈 칸을 다 채웠다면 True 반환
    if index == len(empty_cells):
        return True
    
    # 현재 빈 칸의 좌표 가져오기
    row, col = empty_cells[index]
    
    # 1부터 9까지의 숫자 시도
    for num in range(1, 10):
        if is_valid(board, row, col, num):
            # 유효한 숫자라면 보드에 숫자 배치
            board[row][col] = num
            
            # 다음 빈 칸으로 이동
            if backtrack(board, empty_cells, index + 1):
                return True
            
            # 현재 숫자가 유효하지 않다면 다시 0으로 되돌림
            board[row][col] = 0
    
    # 가능한 숫자가 없다면 False 반환
    return False

def print_board(board):
    # 보드 출력
    for row in board:
        print("".join(map(str, row)))

# 입력 받기
import sys
input = sys.stdin.read

# 입력 데이터 처리
data = input().strip().split()

# 9x9 스도쿠 보드 생성
board = [list(map(int, list(data[i]))) for i in range(9)]

# 스도쿠 풀기
solve_sudoku(board)

# 결과 출력
print_board(board)
